service/history/workflow/mutable_state_impl.go

```diff
@@ -3193,21 +3239,88 @@ func (ms *MutableStateImpl) AddWorkflowExecutionUpdateAcceptedEvent(protocolInst
 		return nil, err
 	}
 	event := ms.hBuilder.AddWorkflowExecutionUpdateAcceptedEvent(protocolInstanceID, updAcceptance)
-	// TODO (alex-update): Async workflow update will require ReplicateWorkflowExecutionUpdateAcceptedEvent
-	// which restores update in the registry if it is not there.
+	if err := ms.ReplicateWorkflowExecutionUpdateAcceptedEvent(event); err != nil {
+		return nil, err
+	}
 	return event, nil
 }
 
+func (ms *MutableStateImpl) ReplicateWorkflowExecutionUpdateAcceptedEvent(
+	event *historypb.HistoryEvent,
+) error {
+	attrs := event.GetWorkflowExecutionUpdateAcceptedEventAttributes()
+	if attrs == nil {
+		return serviceerror.NewInternal("wrong event type in call to ReplicateworkflowExecutionUpdateAcceptedEvent")
+	}
+	ms.updateInfos[attrs.GetAcceptedRequest().GetMeta().GetUpdateId()] = &persistencespb.UpdateInfo{
+		Value: &persistencespb.UpdateInfo_AcceptancePointer{
+			AcceptancePointer: &historyspb.HistoryEventPointer{EventId: event.GetEventId()},
+		},
+	}
+	ms.writeEventToCache(event)
+	return nil
+}
+
 func (ms *MutableStateImpl) AddWorkflowExecutionUpdateCompletedEvent(updResp *updatepb.Response) (*historypb.HistoryEvent, error) {
 	if err := ms.checkMutability(tag.WorkflowActionUpdateCompleted); err != nil {
 		return nil, err
 	}
 	event := ms.hBuilder.AddWorkflowExecutionUpdateCompletedEvent(updResp)
-	// TODO (alex-update): Async workflow update will require ReplicateWorkflowExecutionUpdateCompletedEvent
-	// which removes it from registry and notify update result pollers.
+	if err := ms.ReplicateWorkflowExecutionUpdateCompletedEvent(event); err != nil {
+		return nil, err
+	}
 	return event, nil
 }
 
+func (ms *MutableStateImpl) GetAcceptedWorkflowExecutionUpdates(
```

Comment:
I know there's no caller for this methods so this is more for the next PR.

Do we have any limit on the # of accepted update for a workflow? If no or the limit is large, then I don't think this can work because there can be one history branch read request for each accepted update. 
Even it's like 10 accepted update, it means 10 more read when loading mutable state.

I am also curious to know for an accepted update why do we have to load it's accepted event, the updated request is not used from what I can tell. but maybe I missed something.

Reply:
I've also been discussing the restoration strategy with Alex. It turns out that there's not single point or place in code where it's optimal to restore the full UpdateRegistry state for a given workflow. The next thing I intend to explore is lazily fetching individual update state from MutableState on a per-update basis as messages are processed (i.e., if there is no existing update for a given ID, peek into MS to see if something can be restored before proceeding.

To answer your first question directly: there is a limit to the total number of entries in this index but it's high (2000 has been suggested as the default). 

For the last question - are you suggesting then that rather than storing a pointer into history for updates in the Accepted state, we "inline" that data into the UpdateInfo value so that no history fetch is required?

Reply:
Actually we could store the pointer "just in case" but not dereference it here.